src/mbgl/style/layers/symbol_layer_properties.hpp

```diff
@@ -9,69 +9,251 @@
-    LayoutProperty<bool> iconOptional { false };
-    LayoutProperty<AlignmentType> iconRotationAlignment { AlignmentType::Auto };
-    LayoutProperty<float> iconSize { 1 };
-    LayoutProperty<IconTextFitType> iconTextFit { IconTextFitType::None };
-    LayoutProperty<std::array<float, 4>> iconTextFitPadding { {{ 0, 0, 0, 0 }} };
-    LayoutProperty<std::string> iconImage { "" };
-    LayoutProperty<float> iconRotate { 0 };
-    LayoutProperty<float> iconPadding { 2 };
-    LayoutProperty<bool> iconKeepUpright { false };
-    LayoutProperty<std::array<float, 2>> iconOffset { {{ 0, 0 }} };
-    LayoutProperty<AlignmentType> textPitchAlignment { AlignmentType::Auto };
-    LayoutProperty<AlignmentType> textRotationAlignment { AlignmentType::Auto };
-    LayoutProperty<std::string> textField { "" };
-    LayoutProperty<std::vector<std::string>> textFont { { "Open Sans Regular", "Arial Unicode MS Regular" } };
-    LayoutProperty<float> textSize { 16 };
-    LayoutProperty<float> textMaxWidth { 10 };
-    LayoutProperty<float> textLineHeight { 1.2 };
-    LayoutProperty<float> textLetterSpacing { 0 };
-    LayoutProperty<TextJustifyType> textJustify { TextJustifyType::Center };
-    LayoutProperty<TextAnchorType> textAnchor { TextAnchorType::Center };
-    LayoutProperty<float> textMaxAngle { 45 };
-    LayoutProperty<float> textRotate { 0 };
-    LayoutProperty<float> textPadding { 2 };
-    LayoutProperty<bool> textKeepUpright { true };
-    LayoutProperty<TextTransformType> textTransform { TextTransformType::None };
-    LayoutProperty<std::array<float, 2>> textOffset { {{ 0, 0 }} };
-    LayoutProperty<bool> textAllowOverlap { false };
-    LayoutProperty<bool> textIgnorePlacement { false };
-    LayoutProperty<bool> textOptional { false };
-};
-
-class SymbolPaintProperties {
-public:
-    void cascade(const CascadeParameters&);
-    bool recalculate(const CalculationParameters&);
-
-    PaintProperty<float> iconOpacity { 1 };
-    PaintProperty<Color> iconColor { Color::black() };
-    PaintProperty<Color> iconHaloColor { {} };
-    PaintProperty<float> iconHaloWidth { 0 };
-    PaintProperty<float> iconHaloBlur { 0 };
-    PaintProperty<std::array<float, 2>> iconTranslate { {{ 0, 0 }} };
-    PaintProperty<TranslateAnchorType> iconTranslateAnchor { TranslateAnchorType::Map };
-    PaintProperty<float> textOpacity { 1 };
-    PaintProperty<Color> textColor { Color::black() };
-    PaintProperty<Color> textHaloColor { {} };
-    PaintProperty<float> textHaloWidth { 0 };
-    PaintProperty<float> textHaloBlur { 0 };
-    PaintProperty<std::array<float, 2>> textTranslate { {{ 0, 0 }} };
-    PaintProperty<TranslateAnchorType> textTranslateAnchor { TranslateAnchorType::Map };
+struct SymbolPlacement : LayoutProperty<SymbolPlacementType> {
+    static SymbolPlacementType defaultValue() { return SymbolPlacementType::Point; }
 };
 
+struct SymbolSpacing : LayoutProperty<float> {
+    static float defaultValue() { return 250; }
```

Comment:
Some property types are not `constexpr` (`std::string`, `std:array<float, 2>`). It's possible we could use `constexpr` type variants (`const char *`, `float[2]`) for `defaultValue` but for now I just went with this simple solution. It'll probably be inlined anyway.